core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,475 @@
+      expectedMax: number = expectedNum): void {
+    if (expectedMax === undefined) {
+      expectedMax = expectedNum;
+    }
+    if (args.length < expectedNum || args.length > expectedMax) {
       throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
-    };
+    }
+  }
 
-    var _verifyArgTypesMatchExpectedType = function(argTypes, expectedType) {
-      for (var i = 0; i < argTypes.length; i++) {
-        if (argTypes[i] !== expectedType) {
-          throw new ExprWrongArgTypeError(null, argTypes[i], expectedType);
-        }
+  private verifyArgTypesMatchExpectedType(argTypes: Array<string>,
+      expectedType: string): boolean {
+    for (var i = 0; i < argTypes.length; i++) {
+      if (argTypes[i] !== expectedType) {
+        throw new ExprWrongArgTypeError(null, argTypes[i], expectedType);
       }
-      return true;
-    };
+    }
+    return true;
+  }
 
-    var _verifyArgTypesMatch = function(argType1, argType2) {
-      if (argType1 !== argType2) {
-        throw new ExprWrongArgTypeError(null, argType1, argType2);
-      }
-      return true;
-    };
-
-    var applyFunctionToParseTree = function(parsed, envs, func) {
-      return func(parsed, envs.concat(system));
-    };
-
-    /**
-     * Looks up a variable of the given name in the env. Here the variable
-     * can be system or user defined functions and parameters, as well as
-     * system operators.
-     * @param {string} name The name to look up.
-     * @param {!Array.<!object>} envs Represents a nested name space
-     *     environment to look up the name in. The first element is looked up
-     *     first (i.e. has higher precedence).
-     * @throws {ExprUndefinedVarError} The named variable was not found in
-     *     the given environment.
-     */
-    const lookupEnvs = (name: string, envs: object[]) => {
-      for (const env of envs) {
-        if (env.hasOwnProperty(name)) {
-          return env[name];
-        }
-      }
+  private verifyArgTypesMatch(argType1: string, argType2: string): boolean {
+    if (argType1 !== argType2) {
+      throw new ExprWrongArgTypeError(null, argType1, argType2);
+    }
+    return true;
+  }
 
-      throw new ExprUndefinedVarError(name, envs);
-    };
+  public applyFunctionToParseTree(parsed: string, envs: Array<object>,
+      func: (parsed: string, envs: Array<object>) => string) {
+    return func(parsed, envs.concat(this.system));
+  }
 
-    // Coerces the argument to a Number, and throws an error if the result
-    // is NaN.
-    var _coerceToNumber = function(originalValue) {
-      var coercedValue = (+originalValue);
-      if (!isNaN(coercedValue)) {
-        return coercedValue;
+  /**
+   * Looks up a variable of the given name in the env. Here the variable
+   * can be system or user defined functions and parameters, as well as
+   * system operators.
+   * @param {string} name The name to look up.
+   * @param {!Array.<!object>} envs Represents a nested name space
```

Comment:
Please remove all type information from the JSDoc's, they are no longer useful since typescript does the type checking for us.